### ISSUE TITLE
Internal error in Auto/CaseSplit.hs:284

### ISSUE THREAD

**Issue by brprice** [agda/agda#2261]
> An internal error has occurred. Please report this as a bug.
> Location of the error: `src/full/Agda/Auto/CaseSplit.hs:284`

In the below code, using Auto on the hole results in the above error.
Note that in the `rewrite`, `{_}` is an unsolved metavariable.

``` agda
module InternalError where

data _==_ {l}{A : Set l} : A -> A -> Set l where
  refl : forall {a} -> a == a
{-# BUILTIN EQUALITY _==_ #-}
{-# BUILTIN REFL refl #-}

data D : Set where
  C : D -> D

[_] : D -> D -> D
[ C p ] q = p

lem : {x : D} -> (p q : D) -> [ p ] x == [ q ] x
lem {x} (C p) q rewrite lem {_} p q = {!!} -- Internal error here when C-c C-a
```

**Comment by UlfNorell** [comment]
Here's a version that doesn't require non-standard `BUILTIN` equality:

``` agda
data _==_ {a} {A : Set a} : A → A → Set a where
  refl : ∀ {x} → x == x

data D : Set where
  C : D → D

f : D → D → D
f (C p) q = p

mutual
  X : D
  X = _

  lem : {x : D} → (p q : D) → f p x == f q x
  lem {x} (C p) q with f p X | lem {X} p q
  ... | .(f q X) | refl = {!!} -- Internal error here when C-c C-a
```

**Comment by UlfNorell** [comment]
The internal error arises because a metavariable shows up when auto is not expecting it. Fixing it would require a deeper understanding of the auto code than I believe anyone possesses at the moment.

**Comment by gallais** [comment]
Can't reproduce the internal error (Agsy simply fails to produce anything and finishes in 8ms). It could have already been fixed by #2620.
